database/migrations/0041-update_transactions_for_blockchain.ts

```diff
@@ -0,0 +1,1919 @@
+  // print mysql raw querys
+  interface ReplaceSet {
+    monthName: RegExp
+    monthValue: number
+  }
+  // js month starting with 0 for Jan
+  const replaceSets: ReplaceSet[] = [
+    { monthName: new RegExp('.*Dez.*'), monthValue: -1},
+    { monthName: new RegExp('.*Jan.*'), monthValue: 0 },
+    { monthName: new RegExp('.*Feb.*'), monthValue: 1 },
+    { monthName: new RegExp('.*M.rz.*'), monthValue: 2 },
+    { monthName: new RegExp('.*April.*'), monthValue: 3 },
+  ]
+  const transaction_ids = await queryFn(`
+    SELECT id, balance_date, creation_date, memo 
+    FROM \`transactions\` 
+    WHERE \`balance_date\` = \`creation_date\`
+    AND \`type_id\` = 1
+  `)
+  let downgradeQueries = ''
+  for(let id in transaction_ids) {
+    const transaction = transaction_ids[id]
+    let updatedCreationDate: Date | null = null
+    // determine correct creation date
+    for(const replaceSet of replaceSets) {
+      // check if target creation can be determine which help of the memo
+      if(transaction.memo.match(replaceSet.monthName)) {
+        const oldCreationDate = transaction.creation_date        
+        updatedCreationDate = new Date(oldCreationDate)
+        updatedCreationDate.setMonth(replaceSet.monthValue)
+        break
+      }
+    }
+    // couldn't find info in memo, simply set date 1 month back
+    if(updatedCreationDate === null) {
+      //date.setMonth(date.getMonth() - numOfMonths);
+      updatedCreationDate = new Date(transaction.creation_date)
+      updatedCreationDate.setMonth(transaction.creation_date.getMonth() - 1)
+    }
+    if(updatedCreationDate.getMonth() == 1) {
+      // only 28 februars exist so let us fix it
+      if(updatedCreationDate.getDate() > 28) {
+        updatedCreationDate.setDate(28)
+      }
+    }
+    console.log('// %s, original creation date: %s\nawait queryFn(`UPDATE \\`transactions\\` SET creation_date = \'%s\' WHERE \\`id\\` = %d`)\n', 
+        transaction.memo, 
+        transaction.creation_date.toISOString().slice(0, 19).replace('T', ' '),
+        updatedCreationDate.toISOString().slice(0, 19).replace('T', ' '), 
+        transaction.id
+    )
+    downgradeQueries += 'await queryFn(`UPDATE \\`transactions\\` SET creation_date = \''
+    downgradeQueries += transaction.creation_date.toISOString().slice(0, 19).replace('T', ' ')
+    downgradeQueries += '\' WHERE \\`id\\` = '
+    downgradeQueries += transaction.id
+    downgradeQueries += '`)\n'
+  }
+  console.log("downgrade: \n%s", downgradeQueries)
+  console.log("transactions count: %d", transaction_ids.length) // 224
+*/
+
+  // Erste Schöpfung, viel Spaß damit ;), original creation date: 2019-12-17 13:06:13
+  await queryFn(
+    `UPDATE \`transactions\` SET creation_date = '2019-11-17 13:06:13' WHERE \`id\` = 1`,
+  )
+
+  // Erste Schöpfung, viel Spaß damit ;), original creation date: 2019-12-17 13:06:16
+  await queryFn(
+    `UPDATE \`transactions\` SET creation_date = '2019-11-17 13:06:16' WHERE \`id\` = 2`,
```

Comment:
You habe to correct every following Transaktion for that particular user, since the Transaktion table is considered fixed and unchangeable we cache the data for the final balance in a transaction. If you change the date of an existing transaction you always change the final balance of all following transactions, since the decay will tick more or less depending if you move the date forward or backward. If its before the decay start you still have to adjust the balance of all 3 transactions.

The simplest solution to your problem was the initially proposed one when you raised your validation problem: make an exception:
"Every transaction older or equal to $date can have a maximum of $amount while when newer then $date the amount mussnt be higher then $amount_b"
We do the same with decay:
"If transaction date older then $date do not calculate decay, if newer or equal to $date calculate decay."

Reply:
Decay start time: 2021-05-13 17:46:31-0000
Creation Date of last transaction which will be changed by my migration: 2020-05-03 09:00:07
So none of this transactions are influenced from decay.
But you are right, maybe a follow up transaction cames after the decay, I will check this.